hal-core/src/addr.rs

```diff
@@ -0,0 +1,262 @@
+            self + offset as usize
+        } else {
+            let offset = -offset;
+            self - offset as usize
+        }
+    }
+
+    /// Returns the difference between `self` and `other`.
+    fn difference(self, other: Self) -> isize {
+        if self > other {
+            -(self.as_usize() as isize - other.as_usize() as isize)
+        } else {
+            (other.as_usize() - self.as_usize()) as isize
+        }
+    }
+
+    /// Returns `true` if `self` is aligned on the specified alignment.
+    fn is_aligned<A: Into<usize>>(self, align: A) -> bool {
+        self.align_down(align) == self
+    }
+
+    fn as_ptr(&self) -> *const ();
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+#[repr(transparent)]
+pub struct PAddr(usize);
+
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+#[repr(transparent)]
+pub struct VAddr(usize);
+
+impl fmt::Debug for PAddr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(width) = f.width() {
+            f.debug_tuple("PAddr")
+                .field(&format_args!("{:#0width$x}", self.0, width = width))
+                .finish()
+        } else {
+            f.debug_tuple("PAddr")
+                .field(&format_args!("{:#x}", self.0,))
+                .finish()
+        }
+    }
+}
+
+impl ops::Add<usize> for PAddr {
+    type Output = Self;
+    fn add(self, rhs: usize) -> Self {
+        PAddr(self.0 + rhs)
+    }
+}
+
+impl ops::Add for PAddr {
+    type Output = Self;
+    fn add(self, rhs: Self) -> Self {
+        PAddr(self.0 + rhs.0)
+    }
+}
+
+impl ops::AddAssign for PAddr {
+    fn add_assign(&mut self, rhs: Self) {
+        self.0 += rhs.0;
+    }
+}
+
+impl ops::AddAssign<usize> for PAddr {
+    fn add_assign(&mut self, rhs: usize) {
+        self.0 += rhs;
+    }
+}
+
+impl ops::Sub<usize> for PAddr {
+    type Output = Self;
+    fn sub(self, rhs: usize) -> Self {
+        PAddr(self.0 - rhs)
+    }
+}
+
+impl ops::Sub for PAddr {
+    type Output = Self;
+    fn sub(self, rhs: Self) -> Self {
+        PAddr(self.0 - rhs.0)
+    }
+}
+
+impl ops::SubAssign for PAddr {
+    fn sub_assign(&mut self, rhs: Self) {
+        self.0 -= rhs.0;
+    }
+}
+
+impl ops::SubAssign<usize> for PAddr {
+    fn sub_assign(&mut self, rhs: usize) {
+        self.0 -= rhs;
+    }
+}
+
+impl Address for PAddr {
+    fn as_usize(self) -> usize {
+        self.0 as usize
+    }
+
+    fn align_up<A: Into<usize>>(self, _align: A) -> Self {
+        unimplemented!("eliza")
+    }
+
+    /// Aligns `self` down to `align`.
+    ///
+    /// The specified alignment must be a power of two.
+    fn align_down<A: Into<usize>>(self, _align: A) -> Self {
+        unimplemented!("eliza")
+    }
+
+    /// Returns `true` if `self` is aligned on the specified alignment.
+    fn is_aligned<A: Into<usize>>(self, _align: A) -> bool {
+        unimplemented!("eliza")
+    }
+
+    fn as_ptr(&self) -> *const () {
+        unimplemented!("eliza")
+    }
```

Comment:
Hmm, given that we'll be running with memory virtualisation enabled, it may not make sense to support converting a physical address to a raw pointer, as derefing it will never do what you want.